### PR TITLE
warn about RequireHttpsAttribute with Web APIs

### LINKED ISSUE #5379: Add security warning when using RequireHttpsAttribute  with Web API
For Web API,  do not use RequireHttpsAttribute.  RequireHttpsAttribute  does transparent redirection from HTTP to HTTPS so the client never knows the information was send insecurely over HTTP. Web APIs should close the connection with status code 400 Bad Request and refuse to serve the request.

### PR DESCRIPTION
**Pull request by Rick-Anderson**
Fixes #5379 
[Review URL](https://review.docs.microsoft.com/en-us/aspnet/core/security/enforcing-ssl?branch=pr-en-us-5381)

### CODE REVIEW

File: aspnetcore/security/enforcing-ssl.md [hunk 1]
```diff
@@ -1,37 +1,39 @@
 ---
-title: Enforcing SSL in an ASP.NET Core app
+title: Enforcing TLS in an ASP.NET Core app
 author: rick-anderson
-description: Shows how to require SSL in a ASP.NET Core web app
+description: Shows how to require HTTPS/TSL in a ASP.NET Core web app
```
Comment (guardrex) [inline aspnetcore/security/enforcing-ssl.md]: TLS and recommend a period at the end.

File: aspnetcore/security/enforcing-ssl.md [hunk 2]
```diff
@@ -1,37 +1,39 @@
...
 ms.author: riande
-ms.date: 07/19/2017
+ms.date: 2/9/2018
 ms.prod: asp.net-core
 ms.technology: aspnet
 ms.topic: article
 uid: security/enforcing-ssl
 ---
-# Enforcing SSL in an ASP.NET Core app
+# Enforcing HTTPS/TSL in an ASP.NET Core app
```
Comment (guardrex) [inline aspnetcore/security/enforcing-ssl.md]: TLS

File: aspnetcore/security/enforcing-ssl.md [hunk 3]
```diff
@@ -1,37 +1,39 @@
...
 
 This document shows how to:
 
-- Require SSL for all requests (HTTPS requests only).
+- Require TLS for all requests (HTTPS requests only).
 - Redirect all HTTP requests to HTTPS.
 
-## Require SSL
+> [!WARNING]
+> Do not use [RequireHttpsAttribute](https://docs.microsoft.com/aspnet/core/api/microsoft.aspnetcore.mvc.requirehttpsattribute) on Web APIs that receive sensitive information. `RequireHttpsAttribute` does transparent redirection from HTTP to HTTPS so the client never knows the information was sent insecurely over HTTP. Web APIs should close the connection with status code 400 (Bad Request) and not serve the request.
```
Comment (guardrex) [inline aspnetcore/security/enforcing-ssl.md]: * Do not :point_right: Don't
* The link redirects ... u can use :point_right: `/dotnet/api/Microsoft.AspNetCore.Mvc.RequireHttpsAttribute`
* Comma before "so" in the 2nd sentence

File: aspnetcore/security/enforcing-ssl.md [hunk 4]
```diff
@@ -1,37 +1,39 @@
...
 
-The [RequireHttpsAttribute](https://docs.microsoft.com/aspnet/core/api/microsoft.aspnetcore.mvc.requirehttpsattribute) is used to require SSL. You can decorate controllers or methods with this attribute or you can apply it globally as shown below:
+<a name="require-ssl"></a>
+## Require TLS
 
-Add the following code to `ConfigureServices` in `Startup`:
+The [RequireHttpsAttribute](https://docs.microsoft.com/aspnet/core/api/microsoft.aspnetcore.mvc.requirehttpsattribute) is used to require TSL. You can decorate controllers or methods with this attribute or you can apply it globally. To apply the attribute globally, add the following code to `ConfigureServices` in `Startup`:
```
Comment (guardrex) [inline aspnetcore/security/enforcing-ssl.md]: * Same on link :point_right: `/dotnet/api/Microsoft.AspNetCore.Mvc.RequireHttpsAttribute`
* TSL :point_right: TLS
* Comma before "or" in 2nd sentence

File: aspnetcore/security/enforcing-ssl.md [hunk 5]
```diff
@@ -1,37 +1,39 @@
...
 
 [!code-csharp[Main](authentication/accconfirm/sample/WebApp1/Startup.cs?name=snippet2&highlight=4-999)]
 
-The highlighted code above requires all requests use `HTTPS`, therefore HTTP requests are ignored. The following highlighted code redirects all HTTP requests to HTTPS:
+The preceding highlighted code requires all requests use `HTTPS`, therefore HTTP requests are ignored. The following highlighted code redirects all HTTP requests to HTTPS:
```
Comment (guardrex) [inline aspnetcore/security/enforcing-ssl.md]: therefore :point_right: A semi-colon in front and comma after it. ... or ... Change comma to period and start a new sentence with a comma after "Therefore" in the new sentence.

File: aspnetcore/security/enforcing-ssl.md [hunk 6]
```diff
@@ -1,37 +1,39 @@
...
 
 [!code-csharp[Main](authentication/accconfirm/sample/WebApp1/Startup.cs?name=snippet_AddRedirectToHttps&highlight=7-999)]
 
-See [URL Rewriting Middleware](xref:fundamentals/url-rewriting) for more information.
+ For more information, see [URL Rewriting Middleware](xref:fundamentals/url-rewriting).
 
 Requiring HTTPS globally (`options.Filters.Add(new RequireHttpsAttribute());`) is a security best practice. Applying the 
-`[RequireHttps]` attribute to all controller isn't considered as secure as requiring HTTPS globally. You can't guarantee new controllers added to your app will remember to apply the `[RequireHttps]` attribute.
+`[RequireHttps]` attribute to all controller isn't considered as secure as requiring HTTPS globally. You can't guarantee new controllers/Razor Pages added to your app remember to apply the `[RequireHttps]` attribute.
```
Comment (guardrex) [inline aspnetcore/security/enforcing-ssl.md]: * controller :point_right: controllers (plural)
* 2nd sentence seems awkward .... the "app" remember. How about something along the lines of ...

> You can't guarantee that when new controllers and Razor Pages are added to the app that the developer will remember to apply the `[RequireHttps]` attribute.

**Comment by Rick-Anderson** [comment]
@guardrex  awesome feedback. got rid of future tense and shortened sentence.

File: aspnetcore/security/enforcing-ssl.md [hunk 7]
```diff
@@ -1,7 +1,7 @@
 ---
 title: Enforcing TLS in an ASP.NET Core app
 author: rick-anderson
-description: Shows how to require HTTPS/TSL in a ASP.NET Core web app
+description: Shows how to require HTTPS/TSL in a ASP.NET Core web app.
```
Comment (guardrex) [inline aspnetcore/security/enforcing-ssl.md]: TLS

File: aspnetcore/security/enforcing-ssl.md [hunk 8]
```diff
@@ -1,37 +1,39 @@
...
+<a name="require-ssl"></a>
+## Require TLS
 
-Add the following code to `ConfigureServices` in `Startup`:
+The [RequireHttpsAttribute](/dotnet/api/Microsoft.AspNetCore.Mvc.RequireHttpsAttribute) is used to require TLS. `[RequireHttpsAttribute]` can decorate controllers or methods, or can be applied globally. To apply the attribute globally, add the following code to `ConfigureServices` in `Startup`:
 
 [!code-csharp[Main](authentication/accconfirm/sample/WebApp1/Startup.cs?name=snippet2&highlight=4-999)]
 
-The highlighted code above requires all requests use `HTTPS`, therefore HTTP requests are ignored. The following highlighted code redirects all HTTP requests to HTTPS:
+The preceding highlighted code requires all requests use `HTTPS`; therefore HTTP requests are ignored. The following highlighted code redirects all HTTP requests to HTTPS:
```
Comment (guardrex) [inline aspnetcore/security/enforcing-ssl.md]: still needs a comma after "therefore"

**Comment by guardrex** [comment]
:shipit:

File: aspnetcore/security/enforcing-ssl.md [hunk 9]
```diff
@@ -1,37 +1,39 @@
...
 
 This document shows how to:
 
-- Require SSL for all requests (HTTPS requests only).
+- Require TLS for all requests (HTTPS requests only).
 - Redirect all HTTP requests to HTTPS.
 
-## Require SSL
+> [!WARNING]
+> Do **not** use [RequireHttpsAttribute](/dotnet/api/Microsoft.AspNetCore.Mvc.RequireHttpsAttribute) on Web APIs that receive sensitive information. `RequireHttpsAttribute` does transparent redirection from HTTP to HTTPS, so the client never knows the information was sent insecurely over HTTP. Web APIs should close the connection with status code 400 (Bad Request) and not serve the request.
```
Comment (blowdart) [inline aspnetcore/security/enforcing-ssl.md]: not "transparent". Better would be "RequireHttpsAttribute" uses HTTP status codes to redirect browsers from HTTP to HTTPS which API clients may not understand or obey. Such clients may end up sending information over HTTP. Web APIs should either not listen on HTTP or should close the connection with status code 400 (Bad Request) and not serve the request.

File: aspnetcore/security/enforcing-ssl.md [hunk 10]
```diff
@@ -1,37 +1,39 @@
...
 
-The [RequireHttpsAttribute](https://docs.microsoft.com/aspnet/core/api/microsoft.aspnetcore.mvc.requirehttpsattribute) is used to require SSL. You can decorate controllers or methods with this attribute or you can apply it globally as shown below:
+<a name="require-ssl"></a>
```
Comment (blowdart) [inline aspnetcore/security/enforcing-ssl.md]: require-tls :p

Comment (blowdart) [reply aspnetcore/security/enforcing-ssl.md]: In fact, really require-https

File: aspnetcore/security/enforcing-ssl.md [hunk 11]
```diff
@@ -1,37 +1,39 @@
...
 ms.author: riande
-ms.date: 07/19/2017
+ms.date: 2/9/2018
 ms.prod: asp.net-core
 ms.technology: aspnet
 ms.topic: article
 uid: security/enforcing-ssl
 ---
-# Enforcing SSL in an ASP.NET Core app
+# Enforcing HTTPS/TLS in an ASP.NET Core app
```
Comment (blowdart) [inline aspnetcore/security/enforcing-ssl.md]: Remove /TLS

File: aspnetcore/security/enforcing-ssl.md [hunk 12]
```diff
@@ -1,37 +1,39 @@
...
 
 By [Rick Anderson](https://twitter.com/RickAndMSFT)
 
 This document shows how to:
 
-- Require SSL for all requests (HTTPS requests only).
+- Require TLS for all requests (HTTPS requests only).
```
Comment (blowdart) [inline aspnetcore/security/enforcing-ssl.md]: Require HTTPS for all requests

File: aspnetcore/security/enforcing-ssl.md [hunk 13]
```diff
@@ -1,37 +1,39 @@
...
-## Require SSL
+> [!WARNING]
+> Do **not** use [RequireHttpsAttribute](/dotnet/api/Microsoft.AspNetCore.Mvc.RequireHttpsAttribute) on Web APIs that receive sensitive information. `RequireHttpsAttribute` does transparent redirection from HTTP to HTTPS, so the client never knows the information was sent insecurely over HTTP. Web APIs should close the connection with status code 400 (Bad Request) and not serve the request.
 
-The [RequireHttpsAttribute](https://docs.microsoft.com/aspnet/core/api/microsoft.aspnetcore.mvc.requirehttpsattribute) is used to require SSL. You can decorate controllers or methods with this attribute or you can apply it globally as shown below:
+<a name="require-ssl"></a>
+## Require TLS
 
-Add the following code to `ConfigureServices` in `Startup`:
+The [RequireHttpsAttribute](/dotnet/api/Microsoft.AspNetCore.Mvc.RequireHttpsAttribute) is used to require TLS. `[RequireHttpsAttribute]` can decorate controllers or methods, or can be applied globally. To apply the attribute globally, add the following code to `ConfigureServices` in `Startup`:
```
Comment (blowdart) [inline aspnetcore/security/enforcing-ssl.md]: HTTPS not TLS.

File: aspnetcore/toc.md [hunk 14]
```diff
@@ -305,7 +305,7 @@
 ### [Compatibility](xref:security/data-protection/compatibility/index)
 #### [Share cookies among apps](xref:security/data-protection/compatibility/cookie-sharing)
 #### [Replace <machineKey> in ASP.NET](xref:security/data-protection/compatibility/replacing-machinekey)
-## [Enforce SSL](xref:security/enforcing-ssl)
+## [Enforce TLS](xref:security/enforcing-ssl)
```
Comment (blowdart) [inline aspnetcore/toc.md]: HTTPS not TLS

**Comment by Rick-Anderson** [comment]
@blowdart  slight tweaks of your wording so MT (machine translation) works.

File: aspnetcore/security/enforcing-ssl.md [hunk 15]
```diff
@@ -1,37 +1,41 @@
...
 
 [!code-csharp[Main](authentication/accconfirm/sample/WebApp1/Startup.cs?name=snippet_AddRedirectToHttps&highlight=7-999)]
 
-See [URL Rewriting Middleware](xref:fundamentals/url-rewriting) for more information.
+For more information, see [URL Rewriting Middleware](xref:fundamentals/url-rewriting).
 
-Requiring HTTPS globally (`options.Filters.Add(new RequireHttpsAttribute());`) is a security best practice. Applying the 
-`[RequireHttps]` attribute to all controller isn't considered as secure as requiring HTTPS globally. You can't guarantee new controllers added to your app will remember to apply the `[RequireHttps]` attribute.
+Requiring HTTPS globally (`options.Filters.Add(new RequireHttpsAttribute());`) is a security best practice. Applying the
+`[RequireHttps]` attribute to all controllers/Razor Pages isn't considered as secure as requiring HTTPS globally. You can't guarantee the `[RequireHttps]` attribute is applied when new controllers and Razor Pages are added.
```
Comment (scottaddie) [inline aspnetcore/security/enforcing-ssl.md]: There appears to be 2 spaces after "Applying the".

Comment (Rick-Anderson) [reply aspnetcore/security/enforcing-ssl.md]: That's the deleted code.

**Comment by Rick-Anderson** [comment]
@blowdart  made all your changes. OK to merge?

**Comment by blowdart** [comment]
LGTM

**Pull request merged**